src/common/health.rs

```diff
@@ -0,0 +1,104 @@
+use collection::shards::shard::ShardId;
+use collection::shards::CollectionId;
+use storage::content_manager::toc::TableOfContent;
+use tokio::sync::RwLock;
+
+pub struct Ready {
+    toc: Arc<TableOfContent>,
+    consensus_sync_instant: Instant,
+    initial_unhelathy_shards: RwLock<Option<HashSet<Replica>>>,
+}
+
+impl Ready {
+    pub fn new(toc: Arc<TableOfContent>) -> Self {
+        Self::with_consensus_sync_timeout(toc, Duration::from_secs(10))
+    }
+
+    pub fn with_consensus_sync_timeout(
+        toc: Arc<TableOfContent>,
+        consensus_sync_timeout: Duration,
+    ) -> Self {
+        Self {
+            toc,
+            consensus_sync_instant: Instant::now() + consensus_sync_timeout,
+            initial_unhelathy_shards: Default::default(),
+        }
+    }
+
+    pub async fn ready(&self) -> bool {
+        if Instant::now() < self.consensus_sync_instant {
+            return false;
+        }
+
+        let is_first_run = self.initial_unhelathy_shards.read().await.is_some();
+
+        let this_peer_id = self.toc.this_peer_id;
+        let collections = self.toc.all_collections().await;
+
+        let mut shards = Vec::new();
+
+        for collection in &collections {
+            let state = match self.toc.get_collection(collection).await {
+                Ok(collection) => collection.state().await,
+                Err(_) => continue,
+            };
+
+            for (&shard, info) in state.shards.iter() {
+                if let Some(&state) = info.replicas.get(&this_peer_id) {
+                    shards.push((collection.clone(), shard, state));
+                }
+            }
+        }
+
+        let mut initial_unhealthy_shards = self.initial_unhelathy_shards.write().await;
+
+        match initial_unhealthy_shards.deref_mut() {
+            None => {
+                let unhealthy_shards =
+                    shards.into_iter().filter_map(|(collection, shard, state)| {
+                        if !matches!(state, ReplicaState::Active | ReplicaState::Listener) {
+                            Some(Replica::new(collection, shard))
+                        } else {
+                            None
+                        }
+                    });
+
+                initial_unhealthy_shards
+                    .insert(unhealthy_shards.collect())
+                    .is_empty()
+            }
+
+            Some(initial_unhealthy_shards) => {
+                for (collection, shard, state) in shards {
+                    let replica = Replica::new(collection, shard);
+
+                    if matches!(state, ReplicaState::Active | ReplicaState::Listener) {
+                        initial_unhealthy_shards.remove(&replica);
+                    } else if is_first_run {
+                        initial_unhealthy_shards.insert(replica);
+                    }
+                }
+
+                initial_unhealthy_shards.is_empty()
+            }
+        }
+    }
```

Comment:
I have a few questions regarding `/readyz` implementation.

### First question:

- Do we want `/readyz` to be a continuous check?
  - e.g., we return "non-ready" *any* time *any* shard becomes non-`Active`
- Or do we want `/readyz` to only check for "startup" readiness?
  - e.g., we collect all shards that are non-`Active` after startup and wait till each shard becomes `Active` *once*, and then we always return "ready", even if some shard becomes non-`Active` again

The first two commits () implement "continuous" check, and the last commit () implements "startup" check.

### Second question:

Is read-locking and iterating through `TableOfContent::collections` and `Collection::shards_holder` on each `/readyz` request is fine?

It is possible to make `TableOfContent`/`Collection`/`ShardHolder` *notify* `Ready` when shard changes state and so `Ready` won't have to lock anything... but the code would become, like, 3 times more complicated.

Reply:
> Do we want /readyz to be a continuous check?

I think only for the initial non-active ones. I think its reasonable to assume a node will remain ready even if you do an arbitrary shard move to it, not checking just the initial ones would break this.

> Or do we want /readyz to only check for "startup" readiness?

I think yes.

Maybe Andrey has a different view on this though.